File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.java
import java.util.*;

/*
요금표: 
입차기록
출차기록

[목표]
차량별 주차 요금

*/

class Solution {
    static final String MAX_TIME = "23:59";
    
    static int baseTime, baseFee, unit, unitFee;
    
    public int[] solution(int[] fees, String[] records) {
        baseTime = fees[0];
        baseFee = fees[1];
        unit = fees[2];
        unitFee = fees[3];
        
        Map<String, String> inAt = new HashMap<>();
        Map<String, Integer> feeMap = new HashMap<>();
        Map<String, Integer> accTime = new HashMap<>();
        
        for (int i = 0; i < records.length; i++) {
            String record = records[i];
            
            String[] parts = record.split(" ");
            String time = parts[0];
            String carNum = parts[1];
            String moving = parts[2];
            
            if ("IN".equals(moving)) {
                inAt.put(carNum, time);
            } else {
                String in = inAt.get(carNum);
                String out = time;
                
                int newAccTime = accTime.getOrDefault(carNum, 0) + getDist(in, out);
                accTime.put(carNum, newAccTime);
                
                inAt.remove(carNum);
            }
        }
        
        for (String carNum: inAt.keySet()) {
            String in = inAt.get(carNum);
            
            int spendTime = getDist(in, MAX_TIME);
            int newAccTime = accTime.getOrDefault(carNum, 0) + spendTime;
            
            accTime.put(carNum, newAccTime);
        }
        
        List<String> carNumList = new ArrayList<>(accTime.keySet());
        Collections.sort(carNumList);
        
        int[] answer = new int[carNumList.size()];
        for (int i = 0; i < answer.length; i++) {
            int curAccTime = accTime.get(carNumList.get(i));
            answer[i] = getFee(curAccTime);
        }
        
        return answer;
    }
    
    private int getDist(String in, String out) {
        return Math.abs(normalize(in) - normalize(out));
    }
    
    private int getFee(int spendTime) {
        if (spendTime <= baseTime)
            return baseFee;
        
        int remainTime = spendTime - baseTime;
        int count = remainTime / unit;
        if (remainTime % unit != 0)  
            count++;
        return baseFee + (count) * unitFee;
    }
    
    private int normalize(String time) {
        String[] parts = time.split(":");
        
        int hour = Integer.parseInt(parts[0]);
        int minute = Integer.parseInt(parts[1]);
        
        return 60 * hour + minute;
    }
}

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int toSec(string date) {
    stringstream ss(date);
    
    int h, m;
    char colon;
    
    ss >> h >> colon >> m;
    
    return h * 60 + m;
}

int getTerm(string in, string out) {
    int inSec = toSec(in);
    int outSec = toSec(out);
    
    return abs(outSec - inSec);
}

int calcFee(int total, vector<int>& fees) {
    int base = fees[0];
    int baseFee = fees[1];
    int unitTime = fees[2];
    int unitFee = fees[3];
    
    if (total <= base)
        return baseFee;
    
    int addTime = (total - base);
    
    int cnt = addTime / unitTime;
    if (addTime % unitTime != 0)
        cnt++;
    
    return baseFee + (unitFee * cnt);
}

vector<int> solution(vector<int> fees, vector<string> records) {
    vector<int> answer;
    
    map<string, string> inAt;
    map<string, int> sum;
    
    for (string record: records) {
        stringstream ss(record);
        string word;
        
        vector<string> datas;
        while (ss >> word) {
            datas.push_back(word);
        }
        
        string date = datas[0];
        string car = datas[1];
        string behaviour = datas[2];
        
        if ("IN" == behaviour) {
            inAt[car] = date;
        } else {
            string inDate = inAt[car];
            string outDate = date;
            
            int spendTime = getTerm(inDate, outDate);
            sum[car] = sum[car] + spendTime;
            inAt.erase(car);
        }
    }
    
    for (auto& [car, inDate]: inAt) {
        sum[car] += getTerm(inAt[car], "23:59");
    }
    
    for (auto& [car, time]: sum) {
        int fee = calcFee(time, fees);
        answer.push_back(fee);
    }
    
    return answer;
}

